### ISSUE TITLE
Feature: Inventory menu for grabbing/storing various tools

### ISSUE THREAD

**Issue by Lowie375** [ubc-minetest-classroom/Minetest_Classroom#49]
Currently being developed on [`features/tool-menu`](https://github.com/ubc-minetest-classroom/Minetest_Classroom/tree/features/tool-menu) as a new inventory tab with access to a detached inventory containing various tools.

Current implementation ideas:
1. All players have access to a unified tool menu with a replenishing supply of tools. Each player can see all available tools, but only grab the ones they have the privileges to use
2. Each player has their own tool menu accessible to themselves alone, containing only the tools they have privileges to use

I'm leaning more towards the second implementation, since it would make it less likely for players to acquire tools they should not have access to (if something went wrong).

Development questions:
- Should the menu contain only one copy of each tool (with a "reset" button in case tools get lost), or should it be able to give an infinite amount of copies of a tool (with the option of removing unwanted tools from the inventory)?
  - In other words, should it act like a physical tool belt or a tool "shop"?

**Comment by pauldpickell** [comment]
I agree with implementing an inventory that only displays the tools available to the player based on privs. To avoid confusion, it's probably best to only allow a player to have a single copy of a tool at any given time. This would also prevent a player from spamming the server with unnecessary callbacks and functions. Does this need to be integrated with the tutorial framework based on which level a player has attained?

**Comment by Lowie375** [comment]
This could be integrated with the tutorial framework to gradually give players more space in their toolbox as they earn privileges and gain access to more tools.
We could also have a tutorial which gives players access to the toolbox. This could be used to hide non-essential tools that are unlocked in introductory tutorials away from the player until they reach the "toolbox" tutorial. If we end up including this, I would recommend putting it after the game mechanic tutorials (break-a-block, movement, etc.) and before the other tool-specific tutorials

**Comment by Lowie375** [comment]
Update: a saving/loading inventory has been implemented!

Next task will be to work on a system for restricting which items can be placed into and taken out of the toolbox - likely going to use detached inventory callbacks and check for custom attributes in tool definitions (`_mc_toolmenu_storable = true`)
Alternatively, I could restrict toolbox usage to tool items (in general) instead of tools with a hard-coded attribute. This would be better for scalability, but would also allow the toolbox to be used for *anything*, which may not be a desired behaviour.

**Comment by Lowie375** [comment]
Another issue that needs to be handled is with tool checks:
- Checking if tools are in the player inventory *or* toolbox
- Checking for tool duplication, and removing duplicate tools if present

It may be worth looking into abstracting out parts of the tool give/take framework and integrating that with the toolbox, though that would be a lot more challenging to do with standalone tools such as the magnifying glass in `magnify`.

**Comment by Lowie375** [comment]
The tool handling code has been abstracted out of various tools in the repo and moved to `mc_toolhandler`, which now manages automatic tool giving/taking when a player joins or is granted/revoked a privilege.
Will work on systems to make tool handling more dynamic so that we don't need to hard-code as much information into `mc_toolhandler` and the tools themselves

**Comment by Lowie375** [comment]
Did some digging through the MineTest API, and it seems like it may be possible to move the tool menu directly into the player inventory while still being able to register the necessary callbacks.
This would simplify things a lot, since it means that the toolbox would not need to be saved to mod storage consistently, and that it would be accessed through player inventory calls rather than detached inventory calls. Will prototype + test a refactor and push if it works well.

EDIT: Toolbox has been refactored to use player inventories instead of detached inventories, and all behaviour from the original implementation has been replicated in the new one.

**Comment by pauldpickell** [comment]
One issue that has arisen is if a player deposits tools into the toolbox inventory, leave game, rejoin game, the same tools are issued again to the player's hotbar, and now they have a duplicate set. Perhaps need to add a check in the on_joinplayer callbacks to check if the tool is already in the toolbox.

**Comment by Lowie375** [comment]
Have tested for that behaviour locally and that did not happen. Is this happening with just the forestry tools or all the MineTest Classroom tools?
EDIT (for clarity): if only the forestry tools are affected, that is happening because they are being managed by code within the tools themselves (which only checks the player's `main` inventory). `mc_toolhandler` checks all player inventories (including the toolbox). If this is the case, the fix would be to remove the legacy tool handling code from the tools in `forestry_tools` and add `forestry_tools` to `optional_depends` in `mc_toolhandler/mod.conf`

**Comment by Lowie375** [comment]
> Perhaps need to add a check in the on_joinplayer callbacks to check if the tool is already in the toolbox.

A check for this exists in `mc_toolhandler` - if it is not being run, or is not working as intended, then we have a big problem

**Comment by Lowie375** [comment]
As more tools are added to MineTest Classroom, it may be worth making the toolbox's size scale with the amount of tools available to the player rather than locking it to the default size of 32.
Will begin working on implementing this: current idea is to move the toolbox inventory area into a scrolling area and add some checks to see if the toolbox is full, which will trigger a size increase

**Comment by Lowie375** [comment]
The main implementations of `mc_toolmenu` (the toolbox) and `mc_toolhandler` (the tool management system) are now feature-complete.
Will begin working on an API for ~~`mc_toolmenu`~~ `mc_toolhandler` which can create tool "stores" on a per-player basis so that players can get new copies of the tools they have access to, should the need arise. After that is done, I'll be available to help with any other tasks that need to get done.

**Comment by Lowie375** [comment]
As of commit [`86c3f06`](https://github.com/ubc-minetest-classroom/Minetest_Classroom/commit/86c3f06f94fa05fccf2e9d1b7530004463208f48), feature development of `mc_toolmenu` and `mc_toolhandler` is complete. Future development of these mods will primarily be for bug fixing and/or optimization.
